metric_providers/powermetrics/provider.py

```diff
@@ -35,43 +41,44 @@ def __init__(self, resolution, skip_check=False):
-        if ps.returncode == 1:
-            return False
-        return True
+    def is_our_powermetrics_running(self):
+        pw_processes = subprocess.check_output(['pgrep', '-ix', 'powermetrics'], text=True)
+        total_count = len(pw_processes.strip().split('\n')) if pw_processes else 0
+        minus_startup = total_count -  self._pm_process_count
+        return  minus_startup >= 1
 
 
     def stop_profiling(self):
         try:
             # We try calling the parent method but if this doesn't work we use the more hardcore approach
             super().stop_profiling()
         except PermissionError:
-            #This isn't the nicest way of doing this but there isn't really any other way that is nicer
+            # This isn't the nicest way of doing this but there isn't really any other way that is nicer that doesn't
+            # open up a huge security hole
             subprocess.check_output('sudo /usr/bin/killall powermetrics', shell=True)
             print('Killed powermetrics process with killall!')
+            if self._pm_process_count > 0:
+                print('----------------------------------------------------------------------------------------------')
+                print('This means we will have also killed the already running powermetrics process. Please restart!')
+                print('----------------------------------------------------------------------------------------------')
 
         # As killall returns right after sending the SIGKILL we need to wait and make sure that the process
         # had time to flush everything to disk
-        if self._skip_check:
-            # When this is the case there is probably another powermetrics process running. Which isn't really
-            # that big of a problem but just doesn't work with the way we have implemented this. As the user
-            # has decided to not check anything we will just sleep and assume everything is fine. Not ideal but
-            # will work for now.
-            time.sleep(10)
-        else:
-            count = 0
-            while self.is_powermetrics_running():
-                print(f"Waiting for powermetrics to shut down (try {count}/60). Please do not abort ...")
-                time.sleep(1)
-                count += 1
-                if count >= 60:
-                    subprocess.check_output('sudo /usr/bin/killall -9 powermetrics', shell=True)
-                    raise RuntimeError('powermetrics had to be killed with kill -9. Values can not be trusted!')
+        count = 0
+        while self.is_our_powermetrics_running():
+            print(f"Waiting for powermetrics to shut down (try {count}/60). Please do not abort ...")
+            time.sleep(1)
+            count += 1
+            if count >= 60:
+                subprocess.check_output('sudo /usr/bin/killall -9 powermetrics', shell=True)
+                raise RuntimeError('powermetrics had to be killed with kill -9. Values can not be trusted!')
```

Comment:
Actually what exactly happens in this case? If powermetrics is killed and then a Runtime error is raised the measurement is broken in any case, as no processing is done anymore, right?
Wouldn't it be better to just continue parsing the values and issue a warning that data might be missing as no flusing has occured?

Reply:
I also thought about this. The problem is that we don't really know how much data has actually been flushed. If the process hasn't died after 60 seconds something bad has happened and I would argue the system was in an inconsistent state. So I wanted to fail big and not have any values available at all.